File: ExposureFusion/exposureFusion.cpp
#include <opencv2/photo.hpp>
#include "opencv2/imgcodecs.hpp"
#include <opencv2/highgui.hpp>
#include <vector>
#include <iostream>
#include <fstream>
using namespace cv;
using namespace std;

// Read Images
void readImages(vector<Mat> &images)
{
  
  int numImages = 16;
  static const char* filenames[] =
  {
    "images/memorial0061.jpg",
    "images/memorial0062.jpg",
    "images/memorial0063.jpg",
    "images/memorial0064.jpg",
    "images/memorial0065.jpg",
    "images/memorial0066.jpg",
    "images/memorial0067.jpg",
    "images/memorial0068.jpg",
    "images/memorial0069.jpg",
    "images/memorial0070.jpg",
    "images/memorial0071.jpg",
    "images/memorial0072.jpg",
    "images/memorial0073.jpg",
    "images/memorial0074.jpg",
    "images/memorial0075.jpg",
    "images/memorial0076.jpg"
  };
  
  for(int i=0; i < numImages; i++)
  {
    Mat im = imread(filenames[i]);
    images.push_back(im);
  }

}

int main(int argc, char **argv)
{
  // Read images
  cout << "Reading images ... " << endl;
  vector<Mat> images;
  
  bool needsAlignment = true;
  if(argc > 1)
  {
    // Read images from the command line
    for(int i=1; i < argc; i++)
    {
      Mat im = imread(argv[i]);
      images.push_back(im);
    }

  }
  else
  {
    // Read example images
    readImages(images);
    needsAlignment = false;
  }
  
  // Align input images
  if(needsAlignment)
  {
    cout << "Aligning images ... " << endl;
    Ptr<AlignMTB> alignMTB = createAlignMTB();
    alignMTB->process(images, images);
  }
  else
  {
    cout << "Skipping alignment ... " << endl;
  }
  
  
  // Merge using Exposure Fusion
  cout << "Merging using Exposure Fusion ... " << endl;
  Mat exposureFusion;
  Ptr<MergeMertens> mergeMertens = createMergeMertens();
  mergeMertens->process(images, exposureFusion);
  
  // Save output image
  cout << "Saving output ... exposure-fusion.jpg"<< endl;
  imwrite("exposure-fusion.jpg", exposureFusion * 255);
  
  return EXIT_SUCCESS;
}

File: ExposureFusion/exposureFusion.py
import cv2
import numpy as np
import sys
import os

def readImagesAndTimes(dir_path):
  
  #load images with os
  filenames = [os.path.join(dir_path, f) for f in os.listdir(dir_path) if f.endswith('.jpg')]
  images = []
  for filename in filenames:
    im = cv2.imread(filename)
    im = cv2.resize(im, (500, 500))
    images.append(im)
  
  return images

if __name__ == '__main__':
  
  # Read images
  print("Reading images ... ")
  
  dir_path = '' #path to images
  
  if len(sys.argv) > 1:
    # Read images from the command line
    images = []
    for filename in sys.argv[1:]:
      im = cv2.imread(filename)
      images.append(im)
    needsAlignment = False
  else :
    # Read example images
    images = readImagesAndTimes(dir_path=dir_path)
    needsAlignment = False
  
  # Align input images
  if needsAlignment:
    print("Aligning images ... ")
    alignMTB = cv2.createAlignMTB()
    alignMTB.process(images, images)
  else :
    print("Skipping alignment ... ")
  
  # Merge using Exposure Fusion
  print("Merging using Exposure Fusion ... ");
  mergeMertens = cv2.createMergeMertens()
  exposureFusion = mergeMertens.process(images)

  # Save output image
  print("Saving output ... exposure-fusion.jpg")
  cv2.imwrite("exposure-fusion.jpg", exposureFusion * 255)


